File: server/resolvers.ts
import { createClient } from "@connectrpc/connect";
import { createGrpcTransport } from "@connectrpc/connect-node";
import { UserService, OrganizationService } from "../generated/backend_pb.js";
import type { ExecutionContext, Resolvers } from "./execute.js";

function once(context: ExecutionContext, key: string, load: () => Promise<unknown>) {
  let value = context.cache.get(key);
  if (!value) { value = load(); context.cache.set(key, value); }
  return value;
}

export function createResolvers(userUrl: string, organizationUrl: string): Resolvers {
  const users = createClient(UserService, createGrpcTransport({ baseUrl: userUrl }));
  const organizations = createClient(OrganizationService, createGrpcTransport({ baseUrl: organizationUrl }));
  return {
    "Query.user": (_source, args, context) => once(context, `user:${args.id}`, async () => {
      const response = await users.getUser({ id: String(args.id) }, { signal: context.signal, timeoutMs: 2000 });
      return response.user;
    }),
    "User.organization": (source, _args, context) => {
      if (!source.organizationId) return undefined;
      return once(context, `organization:${source.organizationId}`, async () => {
        const response = await organizations.getOrganization({ id: String(source.organizationId) }, { signal: context.signal, timeoutMs: 2000 });
        return response.organization;
      });
    },
  };
}

File: internal/gateway/resolvers.go
package gateway

import (
	"context"
	"net/http"
	"time"

	"connectrpc.com/connect"
	"github.com/azrsh/fragment-colocation-with-grpc/generated/backendv1"
	"github.com/azrsh/fragment-colocation-with-grpc/generated/backendv1/backendv1connect"
)

func NewResolvers(client *http.Client, userURL, organizationURL string) Resolvers {
	users := backendv1connect.NewUserServiceClient(client, userURL, connect.WithGRPC())
	organizations := backendv1connect.NewOrganizationServiceClient(client, organizationURL, connect.WithGRPC())
	return Resolvers{
		"Query.user": func(_ map[string]any, args map[string]any, ctx *ExecutionContext) (any, error) {
			value, err := scalar(args["id"], "ID", "Query.user(id:)", connect.CodeInvalidArgument)
			if err != nil {
				return nil, err
			}
			id := value.(string)
			return ctx.Once("user:"+id, func() (any, error) {
				requestCtx, cancel := context.WithTimeout(ctx, 2*time.Second)
				defer cancel()
				response, err := users.GetUser(requestCtx, connect.NewRequest(&backendv1.GetUserRequest{Id: id}))
				if err != nil {
					return nil, err
				}
				u := response.Msg.User
				if u == nil {
					return nil, nil
				}
				return map[string]any{"id": u.Id, "name": u.Name, "email": u.Email, "avatarUrl": u.AvatarUrl, "organizationId": u.OrganizationId}, nil
			})
		},
		"User.organization": func(source, args map[string]any, ctx *ExecutionContext) (any, error) {
			id, _ := source["organizationId"].(string)
			if id == "" {
				return nil, nil
			}
			return ctx.Once("organization:"+id, func() (any, error) {
				requestCtx, cancel := context.WithTimeout(ctx, 2*time.Second)
				defer cancel()
				response, err := organizations.GetOrganization(requestCtx, connect.NewRequest(&backendv1.GetOrganizationRequest{Id: id}))
				if err != nil {
					return nil, err
				}
				o := response.Msg.Organization
				if o == nil {
					return nil, nil
				}
				return map[string]any{"id": o.Id, "name": o.Name, "website": o.Website}, nil
			})
		},
	}
}
